### PR TITLE
new lint [`single_call_fn`]

### LINKED ISSUE #10861: `single_call_function`
### What it does

Checks for functions which are only called in 1 location.

Enabling this lint can guard against segmentation which can occur stealthily over a long period of development, when a function is split purposefully for testing or readability this lint can be allowed for these specific cases.

### Advantage

Functions that are only called once, may be better suited as scopes or in-lined.

Premature abstraction can occur in development leading to segmentation with many small functions that are only called once, this can introduce significant additional complexity and indirection that makes the code-base less readable.

### Drawbacks

It can be useful to split a large function into sub-functions even when these are only called once each, this can improve test-ability and readability. This is often done purposefully.

### Example

```rust
fn main() {
    one();
}
fn one() {
    // [...]
}
```

Could be written as:

```rust
fn main() {
    // [...]
}
```

### PR DESCRIPTION
**Pull request by Centri3**
Closes #10861

changelog: New lint [`single_call_fn`]

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
Should we have a test with closures as well?

**Comment by Centri3** [comment]
Good point. It should not lint on those

File: clippy_lints/src/lib.rs [hunk 1]
```diff
@@ -1045,6 +1046,12 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
     });
     let stack_size_threshold = conf.stack_size_threshold;
     store.register_late_pass(move |_| Box::new(large_stack_frames::LargeStackFrames::new(stack_size_threshold)));
+    store.register_late_pass(move |_| {
+        Box::new(single_call_fn::SingleCallFn {
+            avoid_breaking_exported_api,
```
Comment (giraffate) [inline clippy_lints/src/lib.rs:1051]: Can we add this lint to the config file comment?
https://github.com/rust-lang/rust-clippy/blob/cda13a8b26d2d739fe6a4541f5773c8f2f1b3a42/clippy_lints/src/utils/conf.rs#L292-L295

File: clippy_lints/src/single_call_fn.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    fn check_crate_post(&mut self, cx: &LateContext<'tcx>) {
+        let mut v = FnUsageVisitor {
+            cx,
+            def_id_to_usage: &mut self.def_id_to_usage,
+        };
+        cx.tcx.hir().visit_all_item_likes_in_crate(&mut v);
+
+        for usage in self.def_id_to_usage.values() {
+            let fn_span = usage.0;
```
Comment (giraffate) [inline clippy_lints/src/single_call_fn.rs:91]:
```suggestion
            let single_call_fn_span = usage.0;
```

File: clippy_lints/src/single_call_fn.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            if let [usage] = *usage.1 {
```
Comment (giraffate) [inline clippy_lints/src/single_call_fn.rs:92]:
```suggestion
            if let [caller_span] = *usage.1 {
```

**Review by giraffate — COMMENTED** [review]
Overall looks good, thanks! I made small comments.

**Comment by matthiaskrgr** [comment]
Oh we could also use a test showing that this does not lint on `#[test]` functions

**Comment by Centri3** [comment]
Yes, it should not do that and there's currently no check for if it's `#[test]` or not, so I'm pretty sure it will lint

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Comment by Centri3** [comment]
@giraffate, please approve again :D
(I need to get into a habit of doing that...)

**Pull request merged**